### PR TITLE
Local keyring secret store

### PR DESCRIPTION
**Pull request by helbing**
# Description

Implement keyring secret store

## Issue reference

Please reference the issue this PR will close: #1037

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: secretstores/local/keyring/keyring.go [hunk 1]
```diff
@@ -0,0 +1,206 @@
...
+	SecretServiceBackend: true,
+	KeychainBackend:      true,
+	KeyCtlBackend:        true,
+	KWalletBackend:       true,
+	WinCredBackend:       true,
+	FileBackend:          true,
+	PassBackend:          true,
+}
+
+type keyringSecretStoreMetaData struct {
```
Comment (daixiang0) [inline secretstores/local/keyring/keyring.go:52]: Does it mean we can use multiple backends for one component at the same time? Could you share the cases like it?

File: secretstores/local/keyring/keyring.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+	"github.com/dapr/kit/config"
+	"github.com/dapr/kit/logger"
+)
+
+// BackendType is alias of keyring.BackendType.
+type BackendType = keyring.BackendType
+
+// All currently supported secure storage backends.
+const (
+	SecretServiceBackend BackendType = keyring.SecretServiceBackend // secret-service
```
Comment (daixiang0) [inline secretstores/local/keyring/keyring.go:32]: I would not alias here, the old name is meaningful enough.

Comment (helbing) [reply secretstores/local/keyring/keyring.go:32]: Fine, I will change it.

File: secretstores/local/keyring/keyring.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+	KeychainBackend      BackendType = keyring.KeychainBackend      // keychain
+	KeyCtlBackend        BackendType = keyring.KeyCtlBackend        // keyctl
+	KWalletBackend       BackendType = keyring.KWalletBackend       // kwallet
+	WinCredBackend       BackendType = keyring.WinCredBackend       // wincred
+	FileBackend          BackendType = keyring.FileBackend          // file
+	PassBackend          BackendType = keyring.PassBackend          // pass
+)
+
+// availableBackends is a map of backends that keyring secret store are supported.
+var availableBackends = map[BackendType]bool{
```
[the remark this answers is not in the document]
Comment (helbing) [reply secretstores/local/keyring/keyring.go:42]: Just sync with keyring library.

File: secretstores/local/keyring/keyring_test.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		m.Properties = map[string]string{
+			"backendType": string(backendType),
+			"fileDir":     "./",
+		}
+		err := s.Init(m)
+		assert.Nil(t, err)
+	})
+}
+
+func TestGetSecret(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (daixiang0) [reply secretstores/local/keyring/keyring_test.go:71]: We have built and tested for those arch, you can detect OS and run it in the corresponding arch.

For kde we can mark TODO for not. Also you can check tests in the keyring to get some ideas.

[the remark this answers is not in the document]
Comment (helbing) [reply secretstores/local/keyring/keyring_test.go]: Done

File: secretstores/local/keyring/keyring_test.go [hunk 5]
```diff
@@ -0,0 +1,382 @@
+/*
```
Comment (ItalyPaleAle) [inline secretstores/local/keyring/keyring_test.go:1]: Great that you added tests! But, would you consider moving these to be certification tests?

I am concerned with two things:

1. On local machines, if users don't have the dependencies installed (gnupg etc) tests may fail when they run `make test`. We want to make sure that contributors have the best experience even when developing locally, and we should avoid adding dependencies for running unit tests
2. On the CI, installing the dependencies will slow down the test runs for a bit.

These tests are fairly comprehensive already and I think they would be great as certification tests

File: secretstores/local/keyring/keyring.go [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+limitations under the License.
+*/
+
+package keyring
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/99designs/keyring"
```
Comment (ItalyPaleAle) [inline secretstores/local/keyring/keyring.go:20]: It looks like this depends on `github.com/99designs/go-keychain` which requires CGO.

Can you confirm that this PR works **without CGO** (ie. when building and compiling with `CGO_ENABLED=0`) on all 3 platforms? Sadly, using CGO is not possible with Dapr.

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]

File: secretstores/local/keyring/keyring.go (hunk 6 above)
Comment (helbing) [reply secretstores/local/keyring/keyring.go:20]: My fault, I didn't notice that Dapr doesn't support CGO. It seems that except for backends of file and pass,  all require CGO. So this feature, local keyring secret store, should not be needed?

**Comment by ItalyPaleAle** [comment]
@helbing  thanks for the contribution :)

As mentioned in the comment there, we cannot currently accept dependencies that use CGO in Dapr.

We are currently working on a new way to import components so they're out-of-process, and that would allow using CGO as well. You can track the progress here: https://github.com/dapr/dapr/issues/3787

Other proposed components are pending on that, for example SQLite: https://github.com/dapr/components-contrib/pull/1654

**Review by ItalyPaleAle — CHANGES_REQUESTED** [review]
On hold until gRPC components are available

**Comment by daixiang0** [comment]
@helbing @ItalyPaleAle I covert it as draft until gRPC components are available.

**Comment by ItalyPaleAle** [comment]
👋 bot!
Keeping this alive

**Comment by berndverst** [comment]
@helbing take a look at
https://github.com/dapr-sandbox/components-go-sdk
https://docs.dapr.io/developing-applications/develop-components/pluggable-components/

**Comment by ItalyPaleAle** [comment]
I am going to close this PR since it cannot be merged in this repository due to the CGo requirement.

Even with external (pluggable) secret store components, we wouldn't be able to accept this in-tree, as pluggable components are released separately.

Thanks for the contribution and sorry we cannot accept it at this time!

**Pull request closed without merging**